### PR TITLE
[mergify] in order to automate the backport and notifications

### PR DESCRIPTION
**Pull request by v1v**
### What

This is just a proposal to use mergify in order to:

* create the backports automatically
* notify when a backport has a conflict to be resolved.
* notify in the original PR when there is no v\d* labels or backport-skip labels.
* notify when backports are not merged but have been reviewed, this happens once a week.

**NOTE**:

This is a complementary service, so contributors can decide to opt-in and use the labels to automate the backports or use any other kind of process to do the backports

### Why

This is a SaaS that works out of the box with public repositories. So it can simplify your backport.

It's stateless, so it does not track if a particular PR should have been backported, it backports any PR only if:

1) It was merged and tagged with the labels -> https://docs.mergify.io/actions/backport/ and see the `.mergify.yml` file.
2) An explicit GitHub comment was added -> https://docs.mergify.io/commands/#backport

### How does it work?

- Tag, aka add a GitHub label, your PR with `v7.16` or `v7.15` or `backport-skip`
- Then merge your original PR (Pull Request).
- Automatically `mergify` will create the backports and assign the author of the original Pull Request to the new backported PRs.
- If there are any conflicts then, those will be shown in the Description of the backported Pull Request in addition to the `conflicts` github Label. You can resolve those conflicts using the GitHub UI, checking out the PR locally, or as you normally do, if the latter, please close the backported Pull Request.
- By default, the `backport-skip` GitHub label is added in all the new Pull Requests, this helps with the automation. As soon as, the `v7.x` or a similar GitHub label is added the `backport-skip` label is removed.

### What's required?

0) [x] Agree if you'd like to move forward.
1) [ ] Enable mergify (need to ask infra to enable this service)
2) [ ] Agree with the `backport-skip` label since it's required to explicitly comment in the original PR when no backport tags have been added.
3) [ ] Agree with the `v-` or `backport-` labels since product teams use `backport-` mostly.
4) [ ] When a new release branch is created, then it's required to edit `.mergify.yml` to add the new branch backport entry 

### Further details

This particular service has been enabled for different projects:
- https://github.com/elastic/apm-server
- https://github.com/elastic/beats
- https://github.com/elastic/security-docs/

It does allow to auto-merge backported PRs if their GitHub checks passed and some other automations.

### CODE REVIEW

File: .mergify.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+pull_request_rules:
+  - name: ask to resolve conflict
+    conditions:
+      - -merged
+      - -closed
+      - conflict
+    actions:
+        comment:
+          message: |
+            This pull request is now in conflicts. Could you fix it @{{author}}? 🙏
```
Comment (cachedout) [inline .mergify.yml:10]:
```suggestion
            This pull request is now in conflict. Could you fix it @{{author}}? 🙏
```

**Review by cachedout — APPROVED** [review]

**Comment by bmorelli25** [comment]
> Agree with the `backport-skip` label since it's required to explicitly comment in the original PR when no backport tags have been added.

It looks like all of the product repos use this label, so I think it makes sense to match 👍

> Agree with the `v` or `backport-` labels since product teams use `backport-` mostly

@elastic/obs-docs what are your thoughts here? It looks like the beats and kibana repo use `v{version}`, while apm-server uses `backport-{version}`. I don't have a strong opinion one way or the other. I'm tempted to just stick with `v{version}` since that's what we use already.

File: .mergify.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            ```
+            git fetch upstream
+            git checkout -b {{head}} upstream/{{head}}
+            git merge upstream/{{base}}
+            git push upstream {{head}}
+            ```
+  - name: backport patches to 8.0 branch
+    conditions:
+      - merged
+      - base=master
```
Comment (bmorelli25) [inline .mergify.yml:21]: Should these be main? There are a few mentions of `master` in this PR.

```suggestion
      - base=main
```

File: .mergify.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          - "{{ author }}"
+        branches:
+          - "8.0"
+        title: "[{{ destination_branch }}] {{ title }} (backport #{{ number }})"
+        labels:
+          - backport
+  - name: backport patches to 7.16 branch
+    conditions:
+      - merged
+      - base=master
```
Comment (v1v) [inline .mergify.yml:35]:
```suggestion
      - base=main
```

File: .mergify.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          - "{{ author }}"
+        branches:
+          - "7.16"
+        title: "[{{ destination_branch }}] {{ title }} (backport #{{ number }})"
+        labels:
+          - backport
+  - name: backport patches to 7.15 branch
+    conditions:
+      - merged
+      - base=master
```
Comment (v1v) [inline .mergify.yml:49]:
```suggestion
      - base=main
```

File: .mergify.yml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+          - "{{ author }}"
+        branches:
+          - "7.15"
+        title: "[{{ destination_branch }}] {{ title }} (backport #{{ number }})"
+        labels:
+          - backport
+  - name: notify the backport policy
+    conditions:
+      - -label~=(^v\d|backport-skip)
+      - base=master
```
Comment (v1v) [inline .mergify.yml:63]:
```suggestion
      - base=main
```

**Comment by v1v** [comment]
> @elastic/obs-docs what are your thoughts here? It looks like the beats and kibana repo use v{version}, while apm-server uses backport-{version}. I don't have a strong opinion one way or the other. I'm tempted to just stick with v{version} since that's what we use already.

`beats` uses `backport-{version}` while version is 3 digits based
`fleet-server` uses `backport-{version}` while version is 3 digits based
`apm-server` uses `backport-{version}` while version is 2 digits based

**Comment by bmorelli25** [comment]
Ah, if they all use `backport-{version}` then I think we should match. Docs aren't versioned at the patch level, so I'd vote for 2 digit based, e.g., `backport-8.0`.

File: .mergify.yml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            git fetch upstream
+            git checkout -b {{head}} upstream/{{head}}
+            git merge upstream/{{base}}
+            git push upstream {{head}}
+            ```
+  - name: backport patches to 8.0 branch
+    conditions:
+      - merged
+      - base=main
+      - label=v8.0.0
```
Comment (v1v) [inline .mergify.yml:22]:
```suggestion
      - label=backport-8.0
```

File: .mergify.yml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        branches:
+          - "8.0"
+        title: "[{{ destination_branch }}] {{ title }} (backport #{{ number }})"
+        labels:
+          - backport
+  - name: backport patches to 7.16 branch
+    conditions:
+      - merged
+      - base=main
+      - label=v7.16.0
```
Comment (v1v) [inline .mergify.yml:36]:
```suggestion
      - label=backport-7.16
```

File: .mergify.yml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        branches:
+          - "7.16"
+        title: "[{{ destination_branch }}] {{ title }} (backport #{{ number }})"
+        labels:
+          - backport
+  - name: backport patches to 7.15 branch
+    conditions:
+      - merged
+      - base=main
+      - label=v7.15.0
```
Comment (v1v) [inline .mergify.yml:50]:
```suggestion
      - label=backport-7.15
```

File: .mergify.yml [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        assignees:
+          - "{{ author }}"
+        branches:
+          - "7.15"
+        title: "[{{ destination_branch }}] {{ title }} (backport #{{ number }})"
+        labels:
+          - backport
+  - name: notify the backport policy
+    conditions:
+      - -label~=(^v\d|backport-skip)
```
Comment (v1v) [inline .mergify.yml:62]:
```suggestion
      - -label~=^backport
```

File: .mergify.yml [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      - base=main
+    actions:
+      comment:
+        message: |
+          This pull request does not have a backport label. Could you fix it @{{author}}? 🙏
+          To fixup this pull request, you need to add the backport labels for the needed
+          branches, such as:
+          * `v7./d./d` is the label to automatically backport to the `7./d` branch. `/d` is the digit
```
Comment (v1v) [inline .mergify.yml:70]:
```suggestion
          * `backport-/d./d` is the label to automatically backport to the `/d./d` branch. `/d` is the digit
```

File: .mergify.yml [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+          **NOTE**: `backport-skip` has been added to this pull request.
+      label:
+        add:
+          - backport-skip
+  - name: remove backport-skip label
+    conditions:
+      - label~=^v\d
```
Comment (v1v) [inline .mergify.yml:77]:
```suggestion
      - label~=backport-v
      - -merged
      - -closed
```

**Review by bmorelli25 — APPROVED** [review]
Labels LGTM

**Pull request merged**